### DISCUSSION TITLE
Is there a way to get the actual configuration with defaults merged in?

### DISCUSSION THREAD

**Discussion by hyperknot** [ory/kratos#3027] [Q&A]
I'm writing my config for Kratos. I'm looking up every single value from the huge [reference config](https://www.ory.sh/docs/kratos/reference/configuration). My problem is that some values don't mention what is the default. 

Also, I have no idea if `{}` means "turn off everything", or it's the same as if the property was not present.

For example the following:
```
 password:
      config: { }
```

Is this equal to simply deleting password from the config?

Also, does `scope: [ ]` mean empty scope or default scope?

---

Is there any way to get the actual, merged config from Kratos? I mean the one which has everything parsed, env values loaded, default values merged in so that I can actually see what parts of my config are needed?

Some other tools provide this in their --debug output, I don't know if there is something similar to Kratos, either in CLI or via the HTTP interface.

Basically what's `SHOW ALL` in PostgreSQL.

**Comment by vinckr** [comment]
Hello @hyperknot 
A good configuration for getting into Ory Kratos is the [quickstart](https://github.com/ory/kratos/blob/master/contrib/quickstart/kratos/email-password/kratos.yml). The reference configuration documents all options and you usually do not need most of them. Also have a look at the example repo: https://github.com/ory/examples

**Reply by hyperknot** [reply]
Hi @vinckr. I've spent the last 2 days trying basically all config values from quickstart, examples, reference and from the cli `ory get identity config`.

The defaults are not always noted, and my questions regarding empty `[ ]` and `{}` are still valid: how do I know what effect do they have?

For example the scopes are being set to empty array or default?

Hence, it'd be so great to have an option to see what's the actual config, after parsing & merging.

**Reply by hyperknot** [reply]
PostgreSQL `SHOW ALL` is really the best example for this. You can set config values in half a dozen ways, but `SHOW ALL` gives you the final configuration at runtime.

**Reply by vinckr** [reply]
Apologies, I thought you were talking about a configuration for self-hosting Ory Kratos and was confused where you saw empty brackets. 
In Ory Network you have a bunch of settings in the config as placeholders, those are the `[]` and `{}` you are seeing, they don't have an effect. 
So `scope: [ ]` means empty scope, `password: config: { }` means no special configuration is active. 
The config you export through the CLI or load in kratos.yaml is the the one which has everything parsed AFAIK. If you are deploying yourself there are some environmental values for example secret values, but otherwise that should be it.

**Reply by hyperknot** [reply]
I am evaluating both self-hosting and Network. Thanks for clarifying the {} and [].

It's just really not simple to give config values to a black box and trying to guess it's all good. 
For example, reference says that hashing default is bcrypt, while this source code says it's argon2.
https://github.com/ory/kratos/blob/master/driver/config/config.go#L65

But from what I see there isn't really a way to print all the config values out, as they are not stored in a single variable but all the sub-functions are called from different parts off the app.

**Reply by hyperknot** [reply]
OK, I couldn't figure out how does that config.go work. It has
```
DefaultPasswordHashingAlgorithm                          = "argon2"
```
yet the default is actually bcrypt. If I put some print lines in that file they never show in the console so I'm not even sure they are run?

On the other hand, I've read this in the readme:

> Configuration values are now sent to the tracer (e.g. Jaeger) if tracing is enabled.

So there _is_ a way to get the config values out somehow.
